llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp

```diff
@@ -2972,7 +2972,22 @@ bool IRTranslator::translateInsertElement(const User &U,
   Register Res = getOrCreateVReg(U);
   Register Val = getOrCreateVReg(*U.getOperand(0));
   Register Elt = getOrCreateVReg(*U.getOperand(1));
-  Register Idx = getOrCreateVReg(*U.getOperand(2));
+  const auto &TLI = *MF->getSubtarget().getTargetLowering();
+  unsigned PreferredVecIdxWidth = TLI.getVectorIdxTy(*DL).getSizeInBits();
+  Register Idx;
+  if (auto *CI = dyn_cast<ConstantInt>(U.getOperand(2))) {
+    if (CI->getBitWidth() != PreferredVecIdxWidth) {
+      APInt NewIdx = CI->getValue().zextOrTrunc(PreferredVecIdxWidth);
+      auto *NewIdxCI = ConstantInt::get(CI->getContext(), NewIdx);
```

Comment:
I would expect the IRTranslator to pass through the original IR type, and let the legalizer do this casting

Reply:
This is the same as translateExtractElement below. I think it's simpler if all index operands for insert/extract are of size getVectorIdxTy(), no matter where they get created.

Reply:
Then that's also wrong. This is doing legalization work in the IRTranslator

Reply:
The index type in the DAG is not always getVectorIdxTy. The DAG builder passes through the type as-is

Reply:
SelectionDAGBuilder will extend the types as the nodes get constructed in https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp#L3963.
It can be seen in https://godbolt.org/z/eWe1hvjG5.